internal/parse/parse.go

```diff
@@ -331,6 +344,12 @@ func GetCacheMount(args []string, store storage.Store, imageMountLabel string, a
 		if err != nil {
 			return newMount, lockedTargets, fmt.Errorf("unable to change uid,gid of cache directory: %w", err)
 		}
+		if (foundUID || foundGID) && !foundU {
+			if err := chown.ChangeHostPathOwnership(newMount.Source, true, uid, gid); err != nil {
```

Comment:
Does this reset UID to 0 even when I only specify a "gid"?

Reply:
Yes, but I think buildkit does something similar per document here: https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#run---mounttypecache

Reply:
@nalind I tried running this manually on buildkit.

Buildkit Output:
On first invocation i used `uid=` with correct value and RUN step executed successfully and before second invocation of the same file i edited to only keep `gid=` and now i was unable to access file, which hints that the value was reset. But let me know if you want me to try some other use-cases.

#### Before ( first invocation )
```Dockerfile
. . .
RUN --mount=type=cache,uid=100,target=/mycache touch /mycache/go
RUN --mount=type=cache,uid=100,target=/mycache mkdir -p /mycache/dir
RUN --mount=type=cache,uid=100,target=/mycache ls -n /mycache/go
```

```console
. . .
#8 [stage-0 4/6] RUN --mount=type=cache,uid=100,target=/mycache touch /mycache/go
#8 DONE 0.4s

#9 [stage-0 5/6] RUN --mount=type=cache,uid=100,target=/mycache mkdir -p /mycache/dir
#9 DONE 0.4s

#10 [stage-0 6/6] RUN --mount=type=cache,uid=100,target=/mycache ls -n /mycache/go
#10 0.410 -rw-r--r--    1 100      101              0 Aug 11 04:10 /mycache/go
#10 DONE 0.5s
```

#### After ( second invocation )

```Dockerfile
. . .
RUN --mount=type=cache,gid=100,target=/mycache touch /mycache/go
RUN --mount=type=cache,gid=100,target=/mycache mkdir -p /mycache/dir
RUN --mount=type=cache,gid=100,target=/mycache ls -n /mycache/go
```

```console
. . .
#8 [stage-0 4/6] RUN --mount=type=cache,gid=100,target=/mycache touch /mycache/go
#8 0.468 touch: /mycache/go: Permission denied
#8 ERROR: executor failed running [/bin/sh -c touch /mycache/go]: exit code: 1
------
 > [stage-0 4/6] RUN --mount=type=cache,gid=100,target=/mycache touch /mycache/go:
#8 0.468 touch: /mycache/go: Permission denied
------
error: failed to solve: executor failed running [/bin/sh -c touch /mycache/go]: exit code: 1
```

Reply:
The second invocation does indeed appear to be defaulting to creating the cache directory with UID=0 when a UID isn't specified.

AFAIK buildkit doesn't implement changing ownership of the content that's already in a cache mount when a cache mount gets reused, but that's what we're adding.  My question is: what do we do to that content when one of either the UID or GID is not specified?  And is it what a person who hasn't read the implementation would expect?

Reply:
Updated the code to honor `if gid was updated uid stays intact and vice-versa` last sub-test of `build with --mount=type=cache and test ownership` verifies this.